### PR TITLE
Added ability to set file mode on config cache + cache file locking

### PR DESCRIPTION
**Pull request by kynx**
Currently the configuration cache file has permission set to whatever the operating system default is. In many cases this will be world-readable. Since configurations frequently contain sensitive information such as database passwords, this is undesirable.

This PR adds the ability to pass a `$mode` parameter to the constructor (set to 0644 by default for BC) which is then set on the cache file.

For example:

```
$aggregator = new ConfigAggregator([
    MyConfigProvider::class,
], 'data/config-cache.php', 0640);
```

I've also added a (non-blocking) lock when writing to cache. If simultaneous processes try to write to the cache only one will succeed.

### CODE REVIEW

File: src/ConfigAggregator.php [hunk 1]
```diff
@@ -60,19 +60,21 @@ class ConfigAggregator
...
      *     string values representing classes that act as processors. If the
      *     latter, they must be instantiable without constructor arguments.
+     * @param int $mode File mode to set on cached config
      */
     public function __construct(
         array $providers = [],
         $cachedConfigFile = null,
-        array $postProcessors = []
+        array $postProcessors = [],
+        $mode = 0644
```
Comment (Xerkus) [inline src/ConfigAggregator.php]: This is octal 644, which is decimal 420.

Comment (Xerkus) [reply src/ConfigAggregator.php]: There are quite a few setups where server is run as different user, sharing group with cli user. I believe this stricter mode might cause breakage for some flows.

To be safe I would prefer default to be null that will not try to set any mode, keeping current behavior where system umask controls the mode.

File: src/ConfigAggregator.php [hunk 2]
```diff
@@ -60,19 +60,21 @@ class ConfigAggregator
      * @param array $postProcessors Array of processors. These may be callables, or
      *     string values representing classes that act as processors. If the
      *     latter, they must be instantiable without constructor arguments.
+     * @param int $mode File mode to set on cached config
```
Comment (Xerkus) [inline src/ConfigAggregator.php]: I think this extra note is warranted here as using decimal value is a rather subtle mistake.
```suggestion
     * @param int $mode File mode to set on cached config.
     *     Note that mode is an octal value, so to ensure the expected operation,
     *     you need to prefix mode with a zero (eg 0644)
     *     {@see http://php.net/manual/en/function.chmod.php }
```

**Comment by Xerkus** [comment]
I think it would be better off as a part of the configuration rather than a separate constructor argument:
```
$aggregator = new ConfigAggregator(
    [
        new ArrayProvider([
            ConfigAggregator::ENABLE_CACHE => true,
            ConfigAggregator::CACHE_FILEMODE => 0600, // only owner can read and write
        ]),
        new PhpFileProvider('*.global.php'),
    ],
    'data/config-cache.php'
);
```

**Comment by kynx** [comment]
@Xerkus thanks for the review. As suggested, I've move the file mode to the config and kept the current behaviour (where system umask controls the mode) if not present or set to `null`.

**Comment by Xerkus** [comment]
require/include are not respecting advisory locking (no attempt to lock for read is made). That behaviour is correct but it is very surprising and confusing to the unaware.
See https://bugs.php.net/bug.php?id=52895

Some FS might not support `flock()` fully or at all. Eg networking file systems could not support locking between machines or between different OS.
So, instead of acquiring shared lock for `require`-ing cache file, writes should happen to exclusive locked temporary file and which is then moved to achieve atomic update of the actual cache file.

I propose a deterministic cache file name `$cachedConfigFile . '.tmp'` to avoid racing conditions where many processes write their own temporary files en masse, causing load spike when config priming was not used. This is up to discussion as there might be other considerations regarding temporary file selection.

@webimpress, thank you for bringing this to my attention.

**Comment by Xerkus** [comment]
Related zendframework/zend-modulemanager#87

**Comment by kynx** [comment]
> require/include are not respecting advisory locking (no attempt to lock for read is made). That behaviour is correct but it is very surprising and confusing to the unaware.
> See https://bugs.php.net/bug.php?id=52895

Indeed - I hadn't though of that.

> I propose a deterministic cache file name `$cachedConfigFile . '.tmp'` to avoid racing conditions where many processes write their own temporary files en masse, causing load spike when config priming was not used. This is up to discussion as there might be other considerations regarding temporary file selection.

Yeah. If there are a number of instances accessing the cache on a networked filesystem, they won't be aware of each other's lock files.

@Xerkus @webimpress  I'm beginning to think the best approach would be to allow the injection of a PSR-16 SimpleCache, with a simple file-based implementation for BC. That way user's can deal with these problems as they see fit - or switch to using Redis / whatever if they want. I'll raise this on #contributors (once I've extinguished a few fires here at work).

**Comment by kynx** [comment]
Well the simple cache stuff is a bit much for this poor little PR, so I've implemented the temp file + `rename()` as suggested. Hopefully this will address the malformed file issue described in https://github.com/zendframework/zend-modulemanager/pull/88

At the worst, with different instances sharing the code base, each will write the cache file once. 

_However_, `rename()` doesn't work across stream wrappers. So I've had to remove the `vfsStream` stuff from the tests, along with the dev dependency. Hope that's OK.

**Pull request merged**